engines/wintermute/base/gfx/opengl/x/meshx.cpp

```diff
@@ -0,0 +1,486 @@
+			for (int r = 0; r < 4; ++r) {
+				for (int c = 0; c < 4; ++c) {
+					currSkinWeights._offsetMatrix(c, r) = lexer.tokenToFloat();
+					lexer.next();
+					lexer.next();
+				}
+			}
+
+			lexer.next();
+			lexer.next();
+		} else if (lexer.tokenIsIdentifier()) {
+			while (!lexer.reachedClosedBraces()) {
+				lexer.next();
+			}
+
+			lexer.next();
+		} else if (lexer.reachedClosedBraces()) {
+			lexer.next();
+			return true;
+		} else {
+			warning("MeshX::loadFromX unknown token encounterd");
+			lexer.next();
+		}
+	}
+
+	return true;
+}
+
+//////////////////////////////////////////////////////////////////////////
+bool MeshX::generateMesh() {
+	return true;
+}
+
+//////////////////////////////////////////////////////////////////////////
+bool MeshX::findBones(FrameNode *rootFrame) {
+	// normal meshes don't have bones
+	if (!_skinnedMesh) {
+		return true;
+	}
+
+	_boneMatrices.resize(skinWeightsList.size());
+
+	for (uint i = 0; i < skinWeightsList.size(); ++i) {
+		FrameNode *frame = rootFrame->findFrame(skinWeightsList[i]._boneName.c_str());
+
+		if (frame) {
+			_boneMatrices[i] = frame->getCombinedMatrix();
+		} else {
+			warning("MeshX::findBones could not find bone %s", skinWeightsList[i]._boneName.c_str());
+		}
+	}
+
+	return true;
+}
+
+//////////////////////////////////////////////////////////////////////////
+bool MeshX::update(FrameNode *parentFrame) {
+	if (_vertexBuffer == 0) {
+		return false;
+	}
+
+	bool res = false;
+
+	// update skinned mesh
+	if (_skinnedMesh) {
+		BaseArray<Math::Matrix4> finalBoneMatrices;
+		finalBoneMatrices.resize(_boneMatrices.size());
+
+		for (uint i = 0; i < skinWeightsList.size(); ++i) {
+			finalBoneMatrices[i] = *_boneMatrices[i] * skinWeightsList[i]._offsetMatrix;
+		}
+
+		float *vertexDataFloat = reinterpret_cast<float *>(_vertexData);
+
+		for (uint boneIndex = 0; boneIndex < skinWeightsList.size(); ++boneIndex) {
+			for (uint i = 0; i < skinWeightsList[boneIndex]._vertexIndices.size(); ++i) {
+				uint32 vertexIndex = skinWeightsList[boneIndex]._vertexIndices[i];
+				Math::Vector3d pos;
+				pos.setData(_vertexPositionData + vertexIndex * 3);
+				finalBoneMatrices[boneIndex].transform(&pos, true);
+				pos *= skinWeightsList[boneIndex]._vertexWeights[i];
+
+				for (uint j = 0; j < 3; ++j) {
+					vertexDataFloat[vertexIndex * 8 + 5 + j] = pos.getData()[j];
+				}
+			}
+		}
+	} else { // update static mesh
+	}
+	return res;
+}
+
+//////////////////////////////////////////////////////////////////////////
+bool MeshX::updateShadowVol(ShadowVolume *shadow, Math::Matrix4 &modelMat, const Math::Vector3d &light, float extrusionDepth) {
+	if (_vertexBuffer == 0) {
+		return false;
+	}
+
+	return shadow->addMesh(_adjacency, modelMat, light, extrusionDepth);
+}
+
+//////////////////////////////////////////////////////////////////////////
+bool MeshX::render(ModelX *model) {
+	// to be implemented later
+	// Wintermute3D actually two D3DX meshes in the XMesh class

```

Comment:
actually what?

Reply:
I forgot a "has" there. The comment seems irrelevant now anyways, if we are not using VBOs.